php/db.php

```diff
@@ -73,6 +76,103 @@ function GetDataForTable($tabName){
 	return $tabData;
 }
 
+function MigrateDatabase() {
+	global $dbConn, $ip, $userAgent, $dbVersion;
+	$migrationsDir = "SQL/versions/";
+
+
+	StartTimer("MigrateDatabase");
+
+	// Check our current database version
+	$sql = "SELECT config_value FROM config WHERE config_key = 'DATABASE_VERSION';";
+	$data = mysqli_query($dbConn, $sql);
+	
+	$config_result = mysqli_fetch_array($data);
+
+	if ($config_result == NULL) {
+		die("Unable to determine the database version. Please add the config value DATABASE_VERSION before running");
+	}
+
+	$currentDatabaseVersion = intval($config_result[0], 10);
+
+	// Check and see if we need to migrate.
+	if ($dbVersion < $currentDatabaseVersion) {
+		die('Database version is out of date with code version');
+	} elseif ($dbVersion == $currentDatabaseVersion) {
+		StopTimer("MigrateDatabase");
+		return;
+	}
+	
+	// Calculate what migrations exist
+	$migrations = array();
+	
+	$migrationDirFiles = scandir($migrationsDir);
```

Comment:
If I'm understanding this right I think some explicit file name format validation and in-php sorting would be necessary here, as the likes of ".", "..", any folders and any other files which may accidentally match the preg_match below, could be returned here. Also who knows what the OS's ordering will be, or if PHP will even be consistent between versions: will it order ..8, 9, 10, 11... or 1, 10, 100, 101, 102, ...

Normally I wouldn't be so particular, but data is the key of any running 1hgj, and the only thing that can't be recovered, so it needs to be handled super carefully

Reply:
This shouldn't be an issue. The preg_match is narrow enough to not match . or .. or any extraneous files (for instance a README.md which was added to this folder). I will make the preg_match more strict by adding a ^ anchor tag at the beginning.

The OS ordering won't actually cause issues. If you look at lines 115 and 120 you will see that the general method is to actually add the versions to a dictionary style array and loop up from currentDBVersion (defined in database) to targetDBVersion (defined in code) and execute each migration in between in order (if they exist). The OS/php ordering only ever matters if there are duplicate migrations (ie 22_fixLiamsCode.sql and 22_fixMyCode.sql, only one of those will be run). The reason why you don't see a sort() is because the for loop on line 120 orders it for me.

Reply:
I'm adding some extra commenting and working on cleaning this section of code up because I think it is possibly getting misunderstood.